### PR TITLE
Add Gradle build scripts

### LINKED ISSUE #141: Use Gradle for the build
Hi,

I see you have code movement in 8.0.0 branch, so I wonder what do you think of using Gradle as the build system?

It makes lots of things simpler to do (e.g. when compared with Ant and Maven), it and it improves development experience.

I've migrated Apache JMeter from Ant to Gradle, and Apache Calcite from Maven to Gradle.
In both cases, it resulted in significant improvements when it comes to day-to-day development. In both cases, Gradle made it simpler to release new versions.

### PR DESCRIPTION
**Pull request by vlsi**
This is almost feature-complete, so feel free to try it out.

See GitHub Actions CI here: https://github.com/vlsi/javacc/actions

## Overview

This PR implements all the tasks that existed previously (in both Ant and Maven), and the PR adds lots of features:
* Project loads to IDE automatically
* [Reproducible builds](https://reproducible-builds.org/) (the resulting jar, zip, and tar files are the same in case the same Java version is used)
* Build script provides help for the command line
* Code style is managed via [Autostyle](https://github.com/autostyle/autostyle). For now, it trims trailing whitespace. It prints human-understandable messages on failures, and it provides a command to automatically fix violations.
* Release to Central is just a couple of commands (prepareVote + publishDist)
* JavaCC is incremental now. In other words, if a clas (or grammar) is changed, then one can launch `./gradlew test`, and Gradle would build the jar before it starts testing

## CI

GitHub Actions (Linux, macOS, Windows): https://github.com/vlsi/javacc/actions
Travis (Linux): https://travis-ci.com/javacc/javacc/pull_requests

Note: Travis builds with Java 8, and Java 11 fail for some reason.
Java 8 gets blocked at `test/javaFiles` for some reason.
Java 11 fails to download junit3 (it is probably Maven Central or Travis issue)

## Files to be removed as the build is migrated to Gradle

Note: the migration makes `Maven` and some `Ant` files obsolete, so I would recommend to remove them, otherwise it would take significant time to maintain multiple build scripts.

The last commit in the PR removes the files. If you want to compare how Ant or Maven build worked, then skip `Remove redundant build files` commit.

* `**/pom.xml`, `mvn.bat`. Those files are obsolete
* `/build.xml` fully (?) migrated to Gradle + part of it are moved to `examples/build.xml` and `{examples,test}/build.gradle.kts`
* `lib/**`. Dependency management is migrated to Gradle, so `lib` is obsolete

## Testing JavaCC

`./gradlew test` builds the project, and launches all the tests.

## `test` and `examples` folders

I would recommend to eventually migrate the tests to JUnit (or something like that).
Current folder structure makes maintenance really complicated, so I would suggest to execute those folders via Ant (see `./gradlew :javacc-examples:antTest`, and `./gradlew :javacc-test:antTest`)

## Review pom files

`pom.xml` are generated on the fly, and `./gradlew generatePom` can be used to review the pom. It will generate the pom files to `build/publications/*/pom*.xml`

## Release to Central

CI job (validates that publish to Central tasks work): https://github.com/vlsi/vlsi-release-plugins/runs/406605437

Instructions: https://github.com/vlsi/vlsi-release-plugins/tree/master/plugins/stage-vote-release-plugin#making-a-release-candidate

TL;DR:

```
# -Pgh means "push to GitHub + Sonatype"
# if -Pgh is omitted, then it would expect asflike-release-environment and push to localhost

./gradlew prepareVote -Prc=1 -Pgh # prepares a release candidate

./gradlew publishDist -Prc=1 -Pgh # promotes the candidate
```

Note: there's a [playground](https://github.com/vlsi/asflike-release-environment) that launches mock implementations of Git and Nexus in a Docker container, so you can try how release works without pushing to real GitHub / Sonatype servers.

fixes #141

### CODE REVIEW

**Comment by zosrothko** [comment]
I merge your `gradle` branch into my local repo and got build errors: 
[build.log](https://github.com/javacc/javacc/files/4102370/build.log)

**Comment by vlsi** [comment]
@zosrothko , That means `examples\JJTreeExamples\cpp\eg1\history.xml` in your local filesystem has wrong line endings.

(Un)fortunately, Git is not very strict when it comes to the ends of lines that are already present.
Unfortunately, `javacc` repository misses `.gitattribute` file, so there are bad files in the repository (and everybody's workspaces). Git tolerates that, and it just stops complaining after a while.

So the recommended approach is:
1) Merge `.gitattributes` and `.gitignore` changes to the master branch. I mean commits https://github.com/javacc/javacc/pull/143/commits/b56bf4371613fe1da4a8d8947fd31076cca53db4 , https://github.com/javacc/javacc/pull/143/commits/ce9cbaa17d851dc2e1878572749bd0594f0eb1b6 , https://github.com/javacc/javacc/pull/143/commits/2a410e5affa6cbd97caf3f84f2d4c70bbacf0e18

2) Reset workspace so gradle rewrites the local files with their expected line endings.
Note: even if you merge the changes, Git does not overwrite the local files as it `remembers` certain files had improper line endings.
So you need to do something like `git rm --cached -r . && git reset --hard` to re-checkout files.

^^ The above would heal LF vs CRLF issues.


>I merge your gradle branch into my local repo and got build errors

I assume that is because your local files (even after the merge) have the wrong end of line markers.
You can fix that by running the following after the merge.

Note: please ensure you do not have un-committed changes before running commands below.
For instance, if you execute the commands right after the merge then it is fine.

```
git rm --cached -r .
git reset --hard
```

Then the build should succeed.

**Comment by zosrothko** [comment]
gradlew PublishToMavenLocal does not work
```
> Task :publishJavaccPublicationToMavenLocal
Caching disabled for task ':publishJavaccPublicationToMavenLocal' because:
  Build cache is disabled
Task ':publishJavaccPublicationToMavenLocal' is not up-to-date because:
  Task has not declared any outputs despite executing actions.
Publishing to maven local repository
:publishJavaccPublicationToMavenLocal (Thread[Execution worker for ':' Thread 2,5,main]) completed. Took 0.05 secs.
:publishToMavenLocal (Thread[Execution worker for ':' Thread 2,5,main]) started.

> Task :publishToMavenLocal
Skipping task ':publishToMavenLocal' as it has no actions.
:publishToMavenLocal (Thread[Execution worker for ':' Thread 2,5,main]) completed. Took 0.0 secs.
Closing Git repo: C:\Users\fandre\Documents\git\javacc\javacc-7.0\.git
```

**Comment by zosrothko** [comment]
Could you also add all `examples` subdirectories as well as the `tests` one?

**Comment by zosrothko** [comment]
BTW, are you aware of this plugin: [Gradle JavaCC Plugin](https://github.com/johnmartel/javaccPlugin)?

**Comment by vlsi** [comment]
>gradlew PublishToMavenLocal does not work

What do you mean by `does not work` ?

**Comment by vlsi** [comment]
>BTW, are you aware of this plugin: Gradle JavaCC Plugin?

I am not, however relying on the external JavaCC plugins for building JavaCC itself would be extra weird.
If Gradle-JavaCC plugin was located in the same repository, then it could make sense. However, it makes no sense when the plugin is maintained externally.

**Comment by zosrothko** [comment]
> 
> 
> > gradlew PublishToMavenLocal does not work
> 
> What do you mean by `does not work` ?

```
Task :publishToMavenLocal
Skipping task ':publishToMavenLocal' as it has no actions.
:publishToMavenLocal (Thread[Execution worker for ':' Thread 2,5,main]) completed. Took 0.0 secs.
Closing Git repo: C:\Users\fandre\Documents\git\javacc\javacc-7.0\.git
```
I presume skipping a task is not producing any output? (Or did I miss something)

**Comment by zosrothko** [comment]
> I am not, however relying on the external JavaCC plugins for building JavaCC itself would be extra weird.
> If Gradle-JavaCC plugin was located in the same repository, then it could make sense. However, it makes no sense when the plugin is maintained externally.

JavaCC 5 up to 7 rely themself on JavaCC 4.2 (like a bootstrap). And Gradle-JavaCC do not rely on JavaCC itself -- apart running it --. So I do not quite see your point...

**Comment by vlsi** [comment]
>I presume skipping a task is not producing any output? (Or did I miss something)

```
> Task :publishJavaccPublicationToMavenLocal
Caching disabled for task ':publishJavaccPublicationToMavenLocal' because:
  Build cache is disabled
Task ':publishJavaccPublicationToMavenLocal' is not up-to-date because:
  Task has not declared any outputs despite executing actions.
Publishing to maven local repository
:publishJavaccPublicationToMavenLocal (Thread[Execution worker for ':' Thread 2,5,main]) completed. Took 0.05 secs.
:publishToMavenLocal (Thread[Execution worker for ':' Thread 2,5,main]) started.

> Task :publishToMavenLocal
Skipping task ':publishToMavenLocal' as it has no actions.
```

What is says is as follows:
1) Task `publishJavaccPublicationToMavenLocal` was completed successfully (not skipped). And it did publish the artifact to Maven Local repository (you can check that in your `.m2` folder)
2) Task `publishToMavenLocal` is indeed an empty task, and the only thing it does it has a dependency on `publishJavaccPublicationToMavenLocal`. That enables to have human-readable commands (like `gradlew publishToMavenLocal` is much easier for humans than `publishJavaccPublicationToMavenLocal`)
3) `producing any output`. The `output` here means [Gradle definition for `output`](https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:task_inputs_outputs).

You can think of Gradle's tasks as Ant's targets. Sometimes Ant target does nothing but declares dependencies. It does not mean the target `does not work`. Ok?

**Comment by vlsi** [comment]
>Could you also add all examples subdirectories as well as the tests one?

@zosrothko , frankly speaking, it is really sad how the current `examples` and `test` directories are structured in `javacc` repository.

It is extremely sad, that those tests fail with cryptic error messages.
For instance:

```
$ ant
javacc/test/javaFileGeneration/build.xml

test-javafile-contents:

test-static:
     [echo] Static Parser

reset:
   [delete] Deleting directory javacc/test/javaFileGeneration/out-dir
    [mkdir] Created dir: javacc/test/javaFileGeneration/out-dir
     [java] Java Compiler Compiler Version 7.0.6 (Parser Generator)
     [java] (type "javacc" with no arguments for help)
     [java] Reading from file Parser.jj . . .
     [java] File "TokenMgrError.java" does not exist.  Will create one.
     [java] File "ParseException.java" does not exist.  Will create one.
     [java] File "Token.java" does not exist.  Will create one.
     [java] File "SimpleCharStream.java" does not exist.  Will create one.
     [java] Parser generated successfully.

BUILD FAILED
javacc/test/javaFileGeneration/build.xml:34: The following error occurred while executing this line:
javacc/test/javaFileGeneration/build.xml:51: Static ParseException
```

It provides absolutely no clue on what has happened. It makes it hard to debug the test in case something breaks.

That is I would suggest rewriting those tests into proper unit tests that can be executed from IDE.

So far, it makes no sense to migrate the existing folders `tests` and `examples` to Gradle.

**Comment by zosrothko** [comment]
@vlsi First of all, I agree to some extends to your analysis about the examples/tests directories, but that was there since years, and made by Sreeni, the owner of JavaCC. Thus, before changing to anything else, one should be careful to not break anything else. Ok, there are works to do on this subject and may be one will take the challenge to do it.

BTW, could we focus on the gradle build for now. I looked at your Gradle scripts (and buildSrc) and they seem to me far too complex to be maintained for someone without a deep knowledge of Gradle like you, and this is an issue . So I am proposing you to give a try to the Gradle JavaCC plugin in an alternative branch to see if with the same scope of functionnalities (building, testing, deploying, CI, ...), it would be simpler to look at the scripts and understand how they work withotu spending hours.

**Comment by martinswanson** [comment]
Would it be possible to create 1 simple example test with Gradle? Then we can port the rest over at some point in the future. Main thing is to get the plumbing in place.

**Comment by vlsi** [comment]
> I looked at your Gradle scripts (and buildSrc) and they seem to me far too complex to be maintained for someone without a deep knowledge of Gradle like you, and this is an issue

Thanks for looking, however, I would strongly disagree here. The build scripts are much simpler than both Ant and Maven ones. The scripts are WAY easier to support. There's a healthy community (see https://gradle-community.slack.com/ and https://discuss.gradle.org/ ), so even in case you have zero Gradle background, you could just ask questions there, and it is extremely likely you would get support.

Can you please provide the exact items you find to be too complex?
`buildSrc/javacc-bootstrap` 

> So I am proposing you to give a try to the Gradle JavaCC plugin

Do you mean https://github.com/johnmartel/javaccPlugin ?
Why doing that?

Note: `javaccPlugin` results in bad IDE experience. The generated code is not picked up by IDE, so the code is red in the IDE, and so on.
I don't know the history of `javaccPlugin`, however, the plugin does not follow Gradle conventions, so I would refrain from using it at this point.

**Comment by zosrothko** [comment]
@vlsi Could you split your PR in two: one for adding thr build with Gradle and only that, the second with all other fixes. By the way, Sreeni, the owner of JavaCC is  really rejecting all tests based on the generated java files with a reference (mostly this commit https://github.com/javacc/javacc/pull/143/commits/f026df2c472e46699c0f25d570e6a8515911649d). That is why no one has taken the time to update them or even remove them. They were done by a contributor some times ago without any community commitment. So do not spend time in fixing those javaFileGeneration tests since they will be removed.

**Comment by vlsi** [comment]
> Could you split your PR in two: one for adding thr build with Gradle and only that, the second with all other fixes

Done. See https://github.com/javacc/javacc/pull/144
This PR comes on top of #144. The commits are exactly the same.

**Comment by vlsi** [comment]
> That is why no one has taken the time to update them or even remove them

@zosrothko , you or Sereni might like to reject tests, but it turned out, that **this exact test** [helped to find a true bug](https://github.com/javacc/javacc/pull/144/commits/6dd43e978515593a12ea2fd72ad5ea9489a8fc03)

In other words, the current JavaCC parses version as `Double.parseDouble`, and it fails for versions like `7.0.6`.
I don't know the implications, however, it suggests that `parseDouble` should be reworked.

**Comment by vlsi** [comment]
> So do not spend time in fixing those javaFileGeneration tests since they will be removed.

Everything is already fixed or has the appropriate workarounds.

As you see, Travis build passes, [GitHub Actions](https://github.com/vlsi/javacc/actions) passes as well.

**Comment by vlsi** [comment]
@zosrothko , @martinswanson, any feedback here?

**Comment by zosrothko** [comment]
The build of JavaCC with Gradle should use either this plugin [ca.coglinc2.javacc](https://plugins.gradle.org/plugin/ca.coglinc2.javacc) either this plugin [com.intershop.gradle.javacc](https://plugins.gradle.org/plugin/com.intershop.gradle.javacc), the best which matches the requirement of JavaCC: bootstrap with javacc 4.3, build examples for Java, C# and C++ (as in branch 8.0.0), and only Groovy scripts

**Comment by vlsi** [comment]
> and only Groovy scripts

This is a no go for me.

**Comment by vlsi** [comment]
It looks like you have checked exactly none of the items from issue description, but you insist on specific choice for Gradle plugins for no apparent reason.

That is why I see no reason for spending time on PR.

**Comment by zosrothko** [comment]
The reason for using a javacc gradle plugin is to put the Gradle workforce (fix, improvment, knowledge, upgrade and so on) outside the javacc scope. It is the same reason for using javacc-maven-plugin instead of reinventing the wheel. 

Your assumption "I see there's no much interest in Gradle build scripts at this point" is not correct. One has interest in building JavaCC with Gradle but with the requirements as expressed, and in a phased manner. Firest closing, the Maven build, then switching to Gradle.

**Comment by vlsi** [comment]
> The reason for using a javacc gradle plugin is to put the Gradle workforce (fix, improvment, knowledge, upgrade and so on) outside the javacc scope.

That is **very** bad. In practice, JavaCC is extremely unconvenient for integration with build systems. Of course, Gradle and Maven plugins can implement workarounds, however, it would really make sense to fix JavaCC itself so build system plugins could be implemented in a sane way.

**Comment by vlsi** [comment]
> Your assumption "I see there's no much interest in Gradle build scripts at this point" is not correct.

I judge by the number of times different people tried Gradle-based builds, and I judge by the feedback.

Just to recap: my build system does more than all the previous build scripts combined.
It is more powerful and it provides a unified experience.

What do I see in the feedback?
"ca.coglinc2.javacc either this plugin com.intershop.gradle.javacc"?
"and only Groovy scripts"?
"build examples for Java, C# and C++ (as in branch 8.0.0)"?
Really?
You must be kidding.

>One has interest in building JavaCC with Gradle but with the requirements as expressed

The requirements have no technical justification.
On the contrary, you state the requirement of `only Groovy scripts` which is a huge step backward from having a statically-compiled language.

In other words, with Groovy scripts, one never knows which code is acceptable and which is not. It fails in runtime, and you can not predict if it works.

It is similar to the current Ant/Maven scripts. Both `build.xml` and `pom.xml` are just text blobs, and you never know if the syntax is ok or not.

On the other hand, Kotlin is statically typed, so runtime failures are way less often.

>and in a phased manner. Firest closing, the Maven build, then switching to Gradle.

The sooner you drop Maven the better.
Maven will never be able to properly load the project to IDE.

It looks like you never imported JavaCC's sources to IDE. It is fine as long as you are comfortable with it.

However, I am not productive in developing Java code in vim/notepad. I want to have proper code navigation, I want to see cross-references. I want to execute tests one by one and so on.

Gradle makes that seamless.
